src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts

```diff
@@ -283,41 +282,53 @@ export class LocalPythonAndRelatedNonPythonKernelSpecFinder extends LocalKernelS
     }
 
     @capturePerfTelemetry(Telemetry.KernelListingPerf, { kind: 'localPython' })
-    private async listKernelsImplementation(cancelToken: CancellationToken) {
+    private async listKernelsImplementation(cancelToken: CancellationToken, forceRefresh: boolean) {
+        // If we don't have Python extension installed,
+        // then list all of the global python kernel specs.
+        if (!this.extensionChecker.isPythonExtensionInstalled) {
+            await this.appendNewKernels(this.listGlobalPythonKernelSpecs(false));
+            return;
+        }
+
         const interpreters = this.extensionChecker.isPythonExtensionInstalled
             ? this.interpreterService.resolvedEnvironments
             : [];
-
-        traceVerbose(
-            `Listing kernels for ${interpreters.length} interpreters (${interpreters.map((i) => i.id).join(', ')})`
-        );
-        // If we don't have Python extension installed or don't discover any Python interpreters
-        // then list all of the global python kernel specs.
-        if (this.extensionChecker.isPythonExtensionInstalled) {
-            await Promise.all(
-                interpreters.map(async (interpreter) => {
-                    const kernels = await listPythonAndRelatedNonPythonKernelSpecs(
+        await Promise.all(
+            interpreters.map(async (interpreter) => {
+                let finder = this.interpreterKernelSpecs.get(interpreter.id);
+                if (!finder) {
+                    finder = new InterpreterSpecificKernelSpecsFinder(
```

Comment:
Now we find kernelspecs for each interpreter in a seprate class with its own state.
when extension is loading, pytyohn extension would trigger changes to envs every few ms, and that would result in blowing aways everything even though there were not changes to interepters, but only new interpreters were discovoered.

As a result things should be fractionally faster, with fewer resource usage.